src-graphql/client.js

```diff
@@ -22,15 +22,61 @@ export default class Client {
     });
   }
 
+  fetchAll(paginatedModels) {
+    return this.graphQLClient.fetchNextPage(paginatedModels).then(({model}) => {
+      // Until we know how hasNextPage will be exposed, we query until the result is empty
+      if (model.length === 0) {
+        return paginatedModels;
+      }
+
+      paginatedModels.push(...model);
+
+      return this.fetchAll(paginatedModels);
+    });
+  }
+
   fetchAllProducts(query = productConnectionQuery()) {
-    return this.graphQLClient.send(query(this.graphQLClient)).then((response) => {
-      return response.model.shop.products;
+    return this.graphQLClient.send(query(this.graphQLClient)).then(({model, data}) => {
+      const promises = model.shop.products.reduce((promiseAcc, product, i) => {
+        const productData = data.shop.products.edges[i].node;
+
+        // Fetch the rest of the images for this product
+        if (productData.images && productData.images.pageInfo.hasNextPage) {
+          promiseAcc.push(this.fetchAll(product.images));
+        }
+
+        // Fetch the rest of the variants for this product
+        if (productData.variants && productData.variants.pageInfo.hasNextPage) {
+          promiseAcc.push(this.fetchAll(product.variants));
+        }
+
+        return promiseAcc;
+      }, []);
+
+      return Promise.all(promises).then(() => {
+        return model.shop.products;
+      });
     });
   }
 
   fetchProduct(id, query = productQuery()) {
     return this.graphQLClient.send(query(this.graphQLClient, id)).then((response) => {
-      return response.model.node;
+      const promises = [];
+      const productData = response.data.node;
+
+      // Fetch the rest of the images for this product
+      if (productData.images && productData.images.pageInfo.hasNextPage) {
+        promises.push(this.fetchAll(response.model.node.images));
+      }
+
+      // Fetch the rest of the variants for this product
+      if (productData.variants && productData.variants.pageInfo.hasNextPage) {
+        promises.push(this.fetchAll(response.model.node.variants));
+      }
```

Comment:
Since this (L68 - L75) is a shared behaviour between the two product fetching methods, maybe extract it into a `saturateProducts` method or something?